### ISSUE TITLE
No warning when changing array memory address

### ISSUE THREAD

**Issue by scootty1** [numpy/numpy#11107]
I work alot with memory views and from time to time it happens to me that I accidentaly do something which changes the array memory address, like.
For example in my quite extensive program including a memory view at a point before the array has been "fully" defined (for example by appending lines) happens quite often, since there are so many lines of code that I sometimes forget after which point in the code the array unchanged.
What I thought to be a nice addition to numpy to avoid problems like this, was a flag to set to receive a warning/error when trying to perform operations on an array, which could change the memory address and would thus interfere with memory views.
Is that feasible?

**Comment by mattip** [comment]
Isn't this a generic issue with python objects? For instance you can do
```
a = range(3)
b = a
a = range(3) # b and a refer to different objects
```
If both the memory view and your array are in the same namespace, you can check that the data is the same on python3 by checking the `obj` attribute of the memoryview:
```
>>> a = np.arange(3)
>>> b = memoryview(a)
>>> b.obj is a
True
```

**Comment by eric-wieser** [comment]
There's one numpy-specific case I can think of, but it already errors:
```python
>>> a = np.array([1, 2, 3])
>>> b = memoryview(a)
>>> a.resize(5)  # this changes the location of the array in memory
Traceback (most recent call last):
  File "<pyshell#62>", line 1, in <module>
    a.resize(5)
ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function
```

**Comment by scootty1** [comment]
Thanks for your replies. Unfortunately I won't be able to send a full answer before tuesday. But I'll respond asap.

**Comment by scootty1** [comment]
Ok, so I'll try to explain it shortly:
@mattip This does not work for slices of `a`, like in `b = memoryview(a[slice(0, 2)])`.
Or am I wrong?
I am already doing something similar to this with:
```
a = np.arange(20).astype(float)
b = a[slice(5, 9)]
b.__array_interface__['data'][0] == a[slice(5, 9)].__array_interface__['data'][0]
```
But imho this is not a nice way to do it.

@eric-wieser True, but when doing something like `a = np.append(a, 2)` the reference breaks without an error. I am aware of the fact that this is not a numpy specific issue and in fact wanted in alot of cases. But for working with numpy ndarrays something like a flag like `FIXEDMEMORY=True` would be nice, which each function which changes the memory location of an array could check before execution.
I don't know if this is feasible and I also know that such problems can be avoided by good design, but I guess everyone knows that mistakes happen while coding and that such a flag could make life alot easier.

@seberg How would that help? I guess I don't understand enough of numpy to get what you are saying...

**Comment by seberg** [comment]
@scootty1 the thing was, I am confused about what you are doing from the start.

Note that there is always an `arr.base` attribute, just like the `memoryview.obj` one, and they keep pointing upward until you get to whatever actually owns the memory.
There is also the `np.may_share_memory` function which will tell you if there may be a direct ancestry going on.

There is no such thing as "a function that changes the memory location" in normal use cases! `np.append(a, 2)` creates a **new** array, the array that was before called `a` is still very much around (assuming you are using memoryviews that point to the same memory, or slices, etc.).

The numpy functions are very clear in the documentation whether they create a new array (copy) or whether they create views (refer to another array/memoryview) or just change the array in place.

In other words: Under any normal conditions (and it does not sound like you are doing anything crazy), **all arrays have `FIXEDMEMORY==True`** by design.

**Comment by seberg** [comment]
However, because they have fixed memory, that also means that some operations are not possible without creating a new array (such as append). If that is a problem for you, yes, you have to work around it, or avoid these functions (e.g. instead of appending, create an array that is large enough from the start, etc.).

**Comment by mattip** [comment]
as @eric-wieser pointed out,

> b = memoryview(a[slice(0, 2)])

`b.obj.base is a` will give you `True`

**Comment by scootty1** [comment]
@seberg Thanks for your answer!
I've been using `np.shares_memory` for checking the consistency of memory views. But what worked best for me was to save `arr.__array_interface__['data'][0]` of each array as soon as all creating/appending etc. is done, then start my calculations and check for changes in `arr.__array_interface__['data'][0]` afterwards to see if something went wrong (for example I tend to forget the `[:]` when assigning result values to arrays like `arr[:] = some_result_values`). But all these methods require explicit checking. And I thought something like setting a flag could enable implicit checks.

With "a function that changes the memory location" I meant that the memory location to which object `a` is pointing is changing. I know that it will in fact create a new array and that the "old" array will still be around as long as there is still something pointing to its location. I guess I'm not good at using the correct words when describing things like that.

Yep, the numpy documentation is very clear about what is being returned from a function. But what I have trouble with is not the correct use of the numpy functions in general, but rather that I tend to have bugs in my code due to forgetting things like `[:]` or accidentally assigning a memory view before all appending etc. is finished. This happens really seldomly, but it happens from time to time and is always really nasty to debug.
Talking about that topic with colleagues or looking at threads on stackoverflow or other sites, this seems to be happening to most, if not everyone, from time to time. And it seems like alot of people thus tend to avoid using memory views completely, even though they can make things easier and more performant. An easy flag for avoiding problems like this from the start could make memory views alot more attractive.

**Comment by seberg** [comment]
Yeah well... In other words though, this has nothing to do with numpy at all, this is how python works with mutable objects in general.
You have to make sure to actually mutated the mutable object and not replace it, same is true for python lists, though it is probably less common to mutate them in such a way. Btw. using `a[...] = ` is a bit nicer IMO, if you have an N-D array (as oppsed to a known 1-D array).

**Comment by seberg** [comment]
The only thing I can think of, is the use of an IDE/Editor that warns you about redefining a variable name.
Such tools might exist, not sure on top of my head, writing something that lists variables that get redefined should be pretty simple using the AST module.
